### ISSUE TITLE
The values/meaning/colors of the development status field should be configurable per font.

### ISSUE THREAD

**Issue by justvanrossum** [googlefonts/fontra-rcjk#123]
Currently, the list of status codes is hard coded:

https://github.com/googlefonts/fontra-rcjk/blob/bc0b0f2fa3c3cdb81ca39d933ada21e9fd14cbe9/src/fontra_rcjk/base.py#L447-L476

This is modeled after RoboCJK where this is also hard coded:

https://github.com/BlackFoundryCom/robo-cjk/blob/4fc4a7b17b607a47bcd088fd3c6b4f31eeb68b7e/sources/utils/colors.py#L29-L56

This should be fully configurable.

Question one is: to what extent should RoboCJK still support this as well? Or: can we do it in a way that doesn't necessarily break RoboCJK?

If I understand correctly, the django-robocjk level the status field is just an integer, so we can just use arbitrary int values. @fabiocaccamo, can you confirm this?

**Comment by fabiocaccamo** [comment]
@justvanrossum yes, on `django-robocjk` side these colors are [hardcoded](https://github.com/googlefonts/django-robo-cjk/blob/master/robocjk/models.py#L1069-L1073) and mapped to the status field integer value that is stored in the database, this means that we can change colors without touching the database.

Actually changing the colors would have a global effect, there is not a "per-project" or "per-font" colors palette support.

**Comment by justvanrossum** [comment]
Ah hmm I missed that. I’m not sure I understand why the db needs all that information. It does mean we need changes in django-robocjk to implement this feature.

**Comment by fabiocaccamo** [comment]
Colors are needed to recognize the glyphs status when receiving glyphs xml data.

**Comment by justvanrossum** [comment]
I mean, it's *nice* the the admin UI can show the labels, but it is by no means necessary:
<img width="551" alt="image" src="https://github.com/googlefonts/fontra-rcjk/assets/4246121/cff14d74-9ac8-47bc-9cf5-c4a9c10d6883">

**Comment by fabiocaccamo** [comment]
Sorry, I was wrong, because initially we were looking for `public.markColor`, then in a second time we switched to `robocjk.status`.
As you said, status is just an integer and it is recognised as an integer, we have just some mappings (at application level) for showing readable strings in the backend.

**Comment by justvanrossum** [comment]
So we would like the strings and colors to be defined at the font level (in fontLib.json, most likely), and I imagine this can't (easily) be supported by the admin UI. Do you think there may be a better option than "get rid of all the hard coded colors and strings regarding status, and use nothing but an integer"? I mean, I would be fine with that solution, it seems the simplest thing to do.

**Comment by fabiocaccamo** [comment]
If we need to work on this, I would opt for doing the best thing: adding color fields to the font model and read them automatically from `fontLib.json` (with fallback to some defaults).

**Comment by justvanrossum** [comment]
It is not just the label and colors, but also the number of possible status codes. It is currently 5, but could be any number. This shows the form for a definition of a set of statuses:
https://github.com/googlefonts/fontra-rcjk/blob/bc0b0f2fa3c3cdb81ca39d933ada21e9fd14cbe9/src/fontra_rcjk/base.py#L448-L475

Like I said before, I don't really understand why the DB schema would need any more information than "this field is an integer" to begin with.

In Fontra's data model, the status code is stored simply as an integer, and the definitions are only used to present it nicely to the user.

**Comment by fabiocaccamo** [comment]
> It is not just the label and colors, but also the number of possible status codes. It is currently 5, but could be any number. This shows the form for a definition of a set of statuses:

Thank you for the clarification

> Like I said before, I don't really understand why the DB schema would need any more information than "this field is an integer" to begin with.

It doesn't need, it's already just an integer field, we only have labels mapped to integers for presenting the status nicely in backend (I think that reading directly an integer in the backend is not the best thing), but we can remove the labels if needed.

**Comment by justvanrossum** [comment]
> but we can remove the labels if needed

How would the backend currently respond if it receives an unknown/out-of-range status code?

**Comment by fabiocaccamo** [comment]
I think it stores the value even if it's out of range, but in the admin it will show an empty label because unable to map the status code value to an existing label.

Anyway, should try to be 100% sure.

**Comment by justvanrossum** [comment]
Since https://github.com/googlefonts/fontra/issues/1360 is in progress, we need to think how to solve it for django-rcjk.

In my view, the web api should accept any integer number as the status code, and should not make any assumption about the meaning.

Should this become a django-robo-cjk issue?

**Comment by fabiocaccamo** [comment]
@justvanrossum actually if status can't be mapped, it fallbacks to `WIP` status:
https://github.com/googlefonts/django-robo-cjk/blob/master/robocjk/models.py#L1149

**Comment by justvanrossum** [comment]
Ah ok. Then we need to change it so it doesn't do any mapping, just store the integer, whatever value it has.

**Comment by justvanrossum** [comment]
Hm, I think fontra_rcjk only reads and writes the status from/to the glyph's lib data, and doesn't use the API for that at all. So maybe we don't have to do anything.

**Comment by justvanrossum** [comment]
This issue will mostly be fixed by https://github.com/googlefonts/fontra/pull/1389, but we still need to fix reading/writing the status definitions, as this is currently hard-coded.
